### DISCUSSION TITLE
Generated data getting lost when application terminates(pyinstaller bundle-path)

### DISCUSSION THREAD

**Discussion by danishansari** [pyinstaller/pyinstaller#6074] [PyInstaller]
I am generating some images with a binary created using pyinstaller. Since at every run it creates a different bundle path in /tmp.
So I just a create a softlink inside it to map to the my data location. So all the images in the my data location is accessible using the softlink. But when the application terminates it deletes the run time bundle directory inside /tmp which has softlink to the my original data directory, but all the existing files in data directory gets deleted.
Deleting softlink should not delete the original files, so why it is happening?

**Comment by rokm** [comment]
>  Deleting softlink should not delete the original files, so why it is happening?

If you are symlinking a directory inside the temporary bundle directory, then the cleanup code jumps into linked directory and removes all its contents, due to:

https://github.com/pyinstaller/pyinstaller/blob/f159ff969f907184f5600ca9ade88e226fb3c2e6/bootloader/src/pyi_utils.c#L537-L541

The condition should probably be changed into
```
if (S_ISDIR(sbuf.st_mode) && !S_ISLNK(sbuf.st_mode)) { 
```
so that symlinked directories only have the link removed.

That said, surely there are better ways of loading/writing your data (from/to a-priori known external location) in your application than symlinking the data directory into temporary bundle directory? And if not, you could remove the symlink yourself before your python code exits...

**Reply by danishansari** [reply]
That's what I did I unlinked the synlink when program exits, but in worst case it's going to be a disaster for me, coz the data inside the external location is very important and non retrievable.
Thanks for your suggestions, I would love to know different ways of loading the images/videos from external locations.

For context my application is built on Flask server to show images/videos on a webpage. By default Flask can access only static directory to access media.

**Reply by rokm** [reply]
We merged the fix for this problem (#6087), which should be available in upcoming PyInstaller 5.x series.
